warp_drive/training/lightning-trainer.py

```diff
@@ -0,0 +1,520 @@
+
+        # For logging episodic reward
+        self.num_completed_episodes = {}
+        self.episodic_reward_sum = {}
+        self.reward_running_sum = {}
+
+        # Indicates the current timestep of the policy model
+        self.current_timestep = {}
+
+        self.total_steps = self.cuda_envs.episode_length * self.num_episodes
+        self.num_iters = self.total_steps // self.training_batch_size
+
+        for policy in self.policies:
+            self.current_timestep[policy] = 0
+            policy_model_config = self._get_config(["policy", policy, "model"])
+            if policy_model_config["type"] == "fully_connected":
+                self.models[policy] = FullyConnected(
+                    self.cuda_envs,
+                    policy_model_config["fc_dims"],
+                    policy,
+                    self.policy_tag_to_agent_id_map,
+                    create_separate_placeholders_for_each_policy,
+                    obs_dim_corresponding_to_num_agents,
+                )
+            else:
+                raise NotImplementedError
+
+        # Load the model parameters (if model checkpoints are specified)
+        # Note: Loading the model checkpoint may also update the current timestep!
+        self.load_model_checkpoint()
+
+        for policy in self.policies:
+            # Push the models to the GPU
+            self.models[policy].cuda()
+
+            # Initialize the (ADAM) optimizer
+            policy_config = self._get_config(["policy", policy])
+            self.lr_schedules[policy] = ParamScheduler(policy_config["lr"])
+            initial_lr = self.lr_schedules[policy].get_param_value(
+                timestep=self.current_timestep[policy]
+            )
+            self.optimizers[policy] = torch.optim.Adam(
+                self.models[policy].parameters(), lr=initial_lr
+            )
+
+            # Initialize episodic rewards and push to the GPU
+            num_agents_for_policy = len(self.policy_tag_to_agent_id_map[policy])
+            self.num_completed_episodes[policy] = 0
+            self.episodic_reward_sum[policy] = (
+                torch.tensor(0).type(torch.float32).cuda()
+            )
+            self.reward_running_sum[policy] = torch.zeros(
+                (self.num_envs, num_agents_for_policy)
+            ).cuda()
+
+        # Initialize the trainers
+        self.trainers = {}
+        self.clip_grad_norm = {}
+        self.max_grad_norm = {}
+        for policy in self.policies_to_train:
+            algorithm = self._get_config(["policy", policy, "algorithm"])
+            assert algorithm in ["A2C", "PPO"]
+            entropy_coeff = self._get_config(["policy", policy, "entropy_coeff"])
+            vf_loss_coeff = self._get_config(["policy", policy, "vf_loss_coeff"])
+            self.clip_grad_norm[policy] = self._get_config(
+                ["policy", policy, "clip_grad_norm"]
+            )
+            if self.clip_grad_norm[policy]:
```

Comment:
All this logic should be moved to the `configure_gradient_clipping` hook or `clip_gradients`.